extractors/bilibili/bilibili.go

```diff
@@ -55,44 +48,72 @@ func genAPI(aid, cid int, bangumi bool, quality, seasonType, cookie string) (str
 		}
 		utoken = t.Data.Token
 	}
+	var api string
 	if bangumi {
 		// The parameters need to be sorted by name
 		// qn=0 flag makes the CDN address different every time
-		// quality=116(1080P 60) is the highest quality so far
+		// quality=120(4k) is the highest quality so far
 		params = fmt.Sprintf(
-			"appkey=%s&cid=%d&module=bangumi&otype=json&qn=%s&quality=%s&season_type=%s&type=",
-			appKey, cid, quality, quality, seasonType,
+			"avid=%d&cid=%d&bvid=&qn=%d&type=&otype=json&fourk=1&fnver=0&fnval=16",
+			aid, cid, quality,
 		)
 		baseAPIURL = bilibiliBangumiAPI
 	} else {
 		params = fmt.Sprintf(
-			"appkey=%s&cid=%d&otype=json&qn=%s&quality=%s&type=",
-			appKey, cid, quality, quality,
+			"avid=%d&cid=%d&bvid=&qn=%d&type=&otype=json&fourk=1&fnver=0&fnval=16",
+			aid, cid, quality,
 		)
 		baseAPIURL = bilibiliAPI
 	}
+	api = baseAPIURL + params
 	// bangumi utoken also need to put in params to sign, but the ordinary video doesn't need
-	api := fmt.Sprintf(
-		"%s%s&sign=%s", baseAPIURL, params, utils.Md5(params+secKey),
-	)
 	if !bangumi && utoken != "" {
 		api = fmt.Sprintf("%s&utoken=%s", api, utoken)
 	}
 	return api, nil
 }
 
-func genParts(durl []dURLData) ([]*types.Part, int64) {
-	var size int64
-	parts := make([]*types.Part, len(durl))
-	for index, data := range durl {
-		size += data.Size
-		parts[index] = &types.Part{
-			URL:  data.URL,
-			Size: data.Size,
-			Ext:  "flv",
+func genParts(dashData *dashInfo, quality int, referer string) ([]*types.Part, error) {
+	parts := make([]*types.Part, 2)
+	checked := false
+	for _, stream := range dashData.Streams.Video {
+		if stream.ID == quality {
+			s, err := request.Size(stream.BaseURL, referer)
+			if err != nil {
+				return nil, err
+			}
+			parts[0] = &types.Part{
+				URL:  stream.BaseURL,
+				Size: s,
+				Ext:  "mp4",
+			}
+			checked = true
+			break
 		}
 	}
-	return parts, size
+	if !checked {
+		return nil, nil
+	}
+	var audioID int
+	audios := map[int]string{}
+	bandwidth := 0
+	for _, stream := range dashData.Streams.Audio {
+		if stream.Bandwidth > bandwidth {
```

Comment:
I'm not familiar with the new API, what's the difference between these audios?

Reply:
> I'm not familiar with the new API, what's the difference between these audios?

The difference between these audios is the quality, in other words, the bitrate.

Reply:
So we always choose the same audio for different quality videos?

Reply:
Yes. It seems redundant. I will try to move it to so that the program only needs to process once.

Reply:
OK, so far so good, I'm gonna merge this.